test/connectivity/modules/NetworkController/Program.cs

```diff
@@ -0,0 +1,116 @@
+                        await RemoveAllControllingRules(controllers, cts.Token);
+
+                        switch (Settings.Current.NetworkControllerMode)
+                        {
+                            case NetworkControllerMode.OfflineNetworkInterface:
+                                await StartAsync(nic, cts.Token);
+                                break;
+                            case NetworkControllerMode.OfflineTrafficController:
+                                await StartAsync(firewall, cts.Token);
+                                break;
+                            case NetworkControllerMode.SatelliteTrafficController:
+                                await StartAsync(satellite, cts.Token);
+                                break;
+                        }
+                    });
+
+                await cts.Token.WhenCanceled();
+                completed.Set();
+                handler.ForEach(h => GC.KeepAlive(h));
+            }
+            else
+            {
+                Log.LogError($"No network interface found for docker network {Settings.Current.NetworkId}");
+            }
+        }
+
+        static async Task StartAsync(IController controller, CancellationToken cancellationToken)
+        {
+            var delay = Settings.Current.StartAfter;
+
+            INetworkStatusReporter reporter = new NetworkReporter();
+            foreach (Frequency item in Settings.Current.Frequencies)
+            {
+                Log.LogInformation($"Schedule task with {controller.Description} to start after {delay} Offline frequency {item.OfflineFrequency} Online frequency {item.OnlineFrequency} Run times {item.RunsCount}");
+
+                var taskExecutor = new CountedTaskExecutor(
+                    async cs =>
+                    {
+                        await SetNetworkStatus(controller, NetworkStatus.Restricted, reporter, cs);
+                        await Task.Delay(item.OfflineFrequency, cs);
+                        await SetNetworkStatus(controller, NetworkStatus.Default, reporter, cs);
+                    },
+                    delay,
+                    item.OfflineFrequency,
+                    item.RunsCount,
+                    Log,
+                    "restrict/default");
+
+                await taskExecutor.Schedule(cancellationToken);
+
+                // Only needs to set the start delay for first frequency, after that reset to 0
+                delay = TimeSpan.FromSeconds(0);
+            }
+        }
+
+        static async Task SetNetworkStatus(IController controller, NetworkStatus status, INetworkStatusReporter reporter, CancellationToken cs)
+        {
+            await reporter.ReportNetworkStatus(NetworkControllerOperation.SettingRule, status, controller.Description);
+            bool success = await controller.SetStatus(status, cs);
+            await reporter.ReportNetworkStatus(NetworkControllerOperation.RuleSet, status, controller.Description, success);
```

Comment:
What is this method supposed to do?

Reply:
Because there might be a time gap between when the network was set to Offline and when it was reported to the analyzer. Just for direct method call we need to be able to adjust the threshold

---
In reply to: [356452744](https://github.com/Azure/iotedge/pull/2040#discussion_r356452744) [](ancestors = 356452744)